### DISCUSSION TITLE
Composing parsers and decoders, parse overload selection

### DISCUSSION THREAD

**Discussion by kamcma** [pointfreeco/swift-parsing#211] [Q&A] [answered]
Hello.

I am experimenting with trying to compose a pipeline of some existing `Decodable`-based decoding and some `swift-parsing`-based parsing. I looked at the `Conversions.JSON` implementation, but in my scenario, I only have `Decodable` conformance, so I can't implement a bi-directional conversion.

What I've come up with is this:
```swift
extension FooDecodableType {
  static func parser() -> Parsers.AnyParser<Data, Self> {
    .init { (input: inout Data) -> Self in
      try FooDecoder.static.decode(Self.self, from: input)
    }
  }
}

extension BarDecodableType {
  static func parser() -> Parsers.AnyParser<Data, Self> {
    .init { (input: inout Data) -> Self in
      try BarDecoder.static.decode(Self.self, from: input)
    }
  }
}

enum Either {
  case foo(FooDecodableType)
  case bar(BarDecodableType)
}

let eitherParser = OneOf {
  Parse(Either.foo) {
    FooDecodableType.parser()
  }
  Parse(Either.bar) {
    BarDecodableType.parser()
  }
}
```
At this point, I would like to call:
```swift
let myData: Data = // ...
let parsed = try eitherParser.parse(data)
```
I would like to call the following overload, for the ergonomics and because I'm not interested in unconsumed input:
https://github.com/pointfreeco/swift-parsing/blob/493234892f02f461bc9b318381eac103c64b008a/Sources/Parsing/Parser.swift#L43-L67
But Swift selects this overload, I presume because `Data` conforms to `Collection`:
https://github.com/pointfreeco/swift-parsing/blob/493234892f02f461bc9b318381eac103c64b008a/Sources/Parsing/Parser.swift#L69-L136
Processing my input `Data` as a `Collection` like this causes parsing to fail. If I forgo the ergonomic overload, and pass my `Data` as an inout parameter, parsing suceeds:
```swift
var myData: Data = // ...
let parsed = try eitherParser.parse(&data)
```

So, on one level, I suppose I'm asking about overload selection from libraries and if there's a way to elect the overload I want.

On a broader level, though, I wonder what people think about composing decoding and parsing like this, or if maybe there's another way to approach this that would obviate the overload issue. Maybe I should put the decoding in a map parser's `transform` rather than wrapping it in a parser per se, or something? Might need a `Parsers.TryMap` implementation for that.

Thanks for any thoughts!

**Comment by kamcma** [comment]
Thinking on it some more, I am leaning towards decoding in a transform is indeed the idiomatic integration with swift-parsing. `Conversion`'s `apply` and `unapply` are allowed to throw, but `Parsers.Map`'s `transform` cannot, so I'll take a stab at implementing a `Parsers.TryMap` and see if that gets me anywhere.

**Reply by randomeizer** [reply]
I have pitched a variation of `map` that can throw, which @stephencelis and @mbrandonw are not opposed to in concept. Just hasn't made it into the library yet.

All that said, it does seem to fit into transformation, since map, in and of itself, is a one-way operation.

**Comment by stephencelis** [comment, marked as answer]
@kamcma Is there a reason your parsers aren't consuming the `input`? Parsers are supposed to incrementally consume the data handed to them, so you might want to call `input.removeAll()` after successfully decoding those types.

Alternately, you may want to use a conversion here, which better describe non-incremental transformations (and is [what we do with JSON encoding/decoding](https://github.com/pointfreeco/swift-parsing/blob/493234892f02f461bc9b318381eac103c64b008a/Sources/Parsing/Conversions/JSON.swift)). I understand you may not care about creating full parser-printer here, though, and maybe there is no encoder available to use conversions instead. But if you do, you could use the conversion in `MapConversion`.

**Reply by kamcma** [reply]
> @kamcma Is there a reason your parsers aren't consuming the `input`? Parsers are supposed to incrementally consume the data handed to them, so you might want to call `input.removeAll()` after successfully decoding those types.

Just me utterly forgetting the basics while I was experimenting with this 🤦. Thanks.

> Alternately, you may want to use a conversion here, which better describe non-incremental transformations (and is [what we do with JSON encoding/decoding](https://github.com/pointfreeco/swift-parsing/blob/493234892f02f461bc9b318381eac103c64b008a/Sources/Parsing/Conversions/JSON.swift)).

Ah, a distinction between parsing per se being incremental and conversions being non-incremental (if I'm interpreting that right) is clarifying, I think. I think that reinforces that I'll try to do this in a conversion.

Thanks for the pointers.
